packages/next-on-pages/src/buildApplication/processVercelFunctions/edgeFunctions.ts

```diff
@@ -0,0 +1,203 @@
+
+/**
+ * Processes the Edge function routes found in the Vercel build output.
+ *
+ * - Ensures middleware functions have use the correct file for the entrypoint.
+ * - Checks edge function entrypoints exists.
+ * - Tries to fix invalid functions that have a valid edge function that can be used instead.
+ *
+ * @param collectedFunctions Collected functions from the Vercel build output.
+ */
+export async function processEdgeFunctions(
+	collectedFunctions: CollectedFunctions,
+): Promise<void> {
+	const { edgeFunctions, invalidFunctions, ignoredFunctions } =
+		collectedFunctions;
+
+	const tempFunctionsMap = new Map<string, string>();
+	const foundRscFunctions = new Map<string, string>();
+
+	for (const [path, fnInfo] of edgeFunctions) {
+		const { result, finalEntrypoint } = await checkEntrypoint(
+			path,
+			fnInfo.relativePath,
+			fnInfo.config.entrypoint,
+		);
+
+		switch (result) {
+			case 'valid': {
+				// update the edge function with the final entrypoint if it changed
+				if (finalEntrypoint !== fnInfo.config.entrypoint) {
+					fnInfo.config.entrypoint = finalEntrypoint;
+				}
+
+				const formattedPathName = formatRoutePath(fnInfo.relativePath);
+
+				tempFunctionsMap.set(formattedPathName, path);
+				if (formattedPathName.endsWith('/index')) {
+					// strip `/index` from the path name as the build output config doesn't rewrite `/index` to `/`
+					tempFunctionsMap.set(stripIndexRoute(formattedPathName), path);
+				}
+
+				if (formattedPathName.endsWith('.rsc')) {
+					foundRscFunctions.set(formattedPathName, path);
+				}
+				break;
+			}
+			case 'invalid': {
+				invalidFunctions.set(path, fnInfo);
+				edgeFunctions.delete(path);
+				break;
+			}
+			case 'ignore': {
+				ignoredFunctions.set(path, fnInfo);
+				edgeFunctions.delete(path);
+				break;
+			}
+		}
+	}
+
+	// Remove any RSC functions that have a valid non-RSC function that can be used instead.
+	// NOTE: RSC functions are identical to non-RSC functions, so this is okay.
+	// https://github.com/vercel/vercel/blob/fdf86f/packages/next/src/server-build.ts#L1217
+	for (const [formattedPath, path] of foundRscFunctions) {
+		replaceRscWithNonRsc(collectedFunctions, tempFunctionsMap, {
+			formattedPath,
+			path,
+		});
+	}
+
+	await tryToFixInvalidFunctions(collectedFunctions, tempFunctionsMap);
+}
+
+/**
+ * Checks whether the entrypoint file exists for the function.
+ *
+ * If the entrypoint file is `middleware.js`, we check whether the compiled `index.js` exists instead.
+ *
+ * @param fullPath Full path to the function's directory.
+ * @param relativePath Relative path to the function's directory.
+ * @param entrypoint Function entrypoint file name.
+ * @returns Whether the entrypoint is valid, and if the final entrypoint needs to be updated.
+ */
+async function checkEntrypoint(
+	fullPath: string,
+	relativePath: string,
+	entrypoint: string,
+): Promise<
+	| { result: 'ignore' | 'invalid'; finalEntrypoint?: never }
```

Comment:
makes TypeScript happy

![image](https://github.com/cloudflare/next-on-pages/assets/10815538/54050865-7284-44d5-8857-1406cfe3e768)

Reply:
ah sorry I missed that 😓 

that's annoying 😞

I think I'd rather not use destructuring instead of needing to add somewhat unclear typings and do something like this:

![Screenshot 2023-07-20 at 10 47 55](https://github.com/cloudflare/next-on-pages/assets/61631103/e1294908-7ba0-44a6-a777-a6a530275c8b)


But I guess it's a matter of personal preference, I personally don't think that being able to destructure a value is worth the need of using less clear types (as less clear types make the code less clear, not having destructuring doesn't make anything less clear just makes the code slightly less ...modern/convenient?), if you disagree I'm fine with the current code

Reply:
happy to go with that 👍🏻